compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1452,13 +1452,13 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             }));
         }
 
-        for (def_id, res_map) in &tcx.resolutions(()).doc_link_resolutions {
-            record!(self.tables.doc_link_resolutions[def_id.to_def_id()] <- res_map);
-        }
+        //for (def_id, res_map) in &tcx.resolutions(()).doc_link_resolutions {
+        //    record!(self.tables.doc_link_resolutions[def_id.to_def_id()] <- res_map);
+        //}
 
-        for (def_id, traits) in &tcx.resolutions(()).doc_link_traits_in_scope {
-            record_array!(self.tables.doc_link_traits_in_scope[def_id.to_def_id()] <- traits);
-        }
+        //for (def_id, traits) in &tcx.resolutions(()).doc_link_traits_in_scope {
+        //    record_array!(self.tables.doc_link_traits_in_scope[def_id.to_def_id()] <- traits);
+        //}
```

Comment:
@petrochenkov I've been trying to avoid the dependency on `resolutions` by using `doc_link_resolutions` and `doc_link_traits_in_scope` queries. But I could not understand in which case a definition would have an entry in those, and in which case they would ICE.

Is there a general rule?
Query description says "for a module". In which sense? resolve or `mod` items?

Reply:
>Query description says "for a module". In which sense? resolve or `mod` items?

`mod` items (or crate root).
There will be an entry if a module has doc links on items inside it (only exported items or all items, depending on options, resolve/late.rs determines the precise logic).
This is the data that rustdoc currently wants from us.

The query ICEs instead of producing `None` because rustdoc is unreliable, and will have future bugs and unpredictable changes with high probability, this ICE prevents it from going out of sync with the logic in rustc.
Maybe you could use a "private" `None`-returning query used for metadata, and a public ICEing wrapper around it used by rustdoc.